File: services/dotnet-microservices/PaymentService/Infrastructure/Persistence/Repositories/PaymentRepository.cs
namespace PaymentService.Infrastructure.Persistence.Repositories;

using Microsoft.EntityFrameworkCore;
using PaymentService.Domain.Interfaces;
using PaymentService.Domain.Models;
using PaymentService.Infrastructure.Persistence.Entities;

public class PaymentRepository : IPaymentRepository
{
    private readonly AppDbContext _context;

    public PaymentRepository(AppDbContext context) => _context = context;

    public async Task<Payment?> GetByIdAsync(Guid id)
    {
        var entity = await _context.Payments.FindAsync(id);
        return entity == null ? null : MapToDomain(entity);
    }

    public async Task<Payment?> GetByOrderIdAsync(Guid orderId)
    {
        var entity = await _context.Payments.FirstOrDefaultAsync(p => p.OrderId == orderId);
        return entity == null ? null : MapToDomain(entity);
    }

    public async Task<IEnumerable<Payment>> GetByUserIdAsync(Guid userId)
    {
        var entities = await _context.Payments.Where(p => p.UserId == userId).ToListAsync();
        return entities.Select(MapToDomain);
    }

    public async Task<Payment> CreateAsync(Payment payment)
    {
        var entity = MapToEntity(payment);
        _context.Payments.Add(entity);
        await _context.SaveChangesAsync();
        return MapToDomain(entity);
    }

    public async Task<Payment> UpdateAsync(Payment payment)
    {
        var entity = await _context.Payments.FindAsync(payment.Id)
            ?? throw new KeyNotFoundException($"Payment {payment.Id} not found");
        entity.Status = payment.Status;
        entity.TransactionId = payment.TransactionId;
        entity.FailureReason = payment.FailureReason;
        entity.UpdatedAt = payment.UpdatedAt;
        await _context.SaveChangesAsync();
        return MapToDomain(entity);
    }

    public async Task<Refund> CreateRefundAsync(Refund refund)
    {
        var entity = new RefundEntity
        {
            Id = refund.Id, PaymentId = refund.PaymentId, Amount = refund.Amount,
            Reason = refund.Reason, Status = refund.Status, CreatedAt = refund.CreatedAt
        };
        _context.Refunds.Add(entity);
        await _context.SaveChangesAsync();
        return new Refund(entity.Id, entity.PaymentId, entity.Amount, entity.Reason, entity.Status, entity.CreatedAt);
    }

    public async Task<IEnumerable<Refund>> GetRefundsByPaymentIdAsync(Guid paymentId)
    {
        var entities = await _context.Refunds.Where(r => r.PaymentId == paymentId).ToListAsync();
        return entities.Select(e => new Refund(e.Id, e.PaymentId, e.Amount, e.Reason, e.Status, e.CreatedAt));
    }

    private static Payment MapToDomain(PaymentEntity e) => new(
        e.Id, e.OrderId, e.UserId, e.Amount, e.Currency, e.Status,
        e.PaymentMethod, e.TransactionId, e.FailureReason, e.CreatedAt, e.UpdatedAt
    );

    private static PaymentEntity MapToEntity(Payment p) => new()
    {
        Id = p.Id, OrderId = p.OrderId, UserId = p.UserId, Amount = p.Amount,
        Currency = p.Currency, Status = p.Status, PaymentMethod = p.PaymentMethod,
        TransactionId = p.TransactionId, FailureReason = p.FailureReason,
        CreatedAt = p.CreatedAt, UpdatedAt = p.UpdatedAt
    };
}

File: services/nodejs-microservices/payment-service/src/infrastructure/persistence/PaymentRepository.ts
import { pool } from './database';
import { Payment, PaymentStatus, PaymentMethod, Refund, RefundStatus } from '../../domain/models/Payment';

export class PaymentRepository {
  async create(payment: Omit<Payment, 'createdAt' | 'updatedAt'>): Promise<Payment> {
    const query = `
      INSERT INTO payments (id, order_id, user_id, amount, currency, method, status, provider_transaction_id, failure_reason)
      VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9)
      RETURNING *
    `;
    const result = await pool.query(query, [
      payment.id, payment.orderId, payment.userId, payment.amount,
      payment.currency, payment.method, payment.status,
      payment.providerTransactionId, payment.failureReason,
    ]);
    return this.mapToPayment(result.rows[0]);
  }

  async findById(id: string): Promise<Payment | null> {
    const result = await pool.query('SELECT * FROM payments WHERE id = $1', [id]);
    return result.rows.length > 0 ? this.mapToPayment(result.rows[0]) : null;
  }

  async findByOrderId(orderId: string): Promise<Payment | null> {
    const result = await pool.query('SELECT * FROM payments WHERE order_id = $1 ORDER BY created_at DESC LIMIT 1', [orderId]);
    return result.rows.length > 0 ? this.mapToPayment(result.rows[0]) : null;
  }

  async updateStatus(id: string, status: PaymentStatus, transactionId?: string, failureReason?: string): Promise<Payment | null> {
    const query = `
      UPDATE payments SET status = $1, provider_transaction_id = COALESCE($2, provider_transaction_id),
      failure_reason = COALESCE($3, failure_reason), updated_at = CURRENT_TIMESTAMP
      WHERE id = $4 RETURNING *
    `;
    const result = await pool.query(query, [status, transactionId, failureReason, id]);
    return result.rows.length > 0 ? this.mapToPayment(result.rows[0]) : null;
  }

  async createRefund(refund: Omit<Refund, 'createdAt'>): Promise<Refund> {
    const query = `
      INSERT INTO refunds (id, payment_id, amount, reason, status)
      VALUES ($1, $2, $3, $4, $5)
      RETURNING *
    `;
    const result = await pool.query(query, [refund.id, refund.paymentId, refund.amount, refund.reason, refund.status]);
    return this.mapToRefund(result.rows[0]);
  }

  private mapToPayment(row: any): Payment {
    return {
      id: row.id, orderId: row.order_id, userId: row.user_id,
      amount: parseFloat(row.amount), currency: row.currency,
      method: row.method as PaymentMethod, status: row.status as PaymentStatus,
      providerTransactionId: row.provider_transaction_id,
      failureReason: row.failure_reason,
      createdAt: row.created_at, updatedAt: row.updated_at,
    };
  }

  private mapToRefund(row: any): Refund {
    return {
      id: row.id, paymentId: row.payment_id, amount: parseFloat(row.amount),
      reason: row.reason, status: row.status as RefundStatus, createdAt: row.created_at,
    };
  }
}
